### PR TITLE
Fix getAccountProof for Bonsai

### PR DESCRIPTION
**Pull request by gfukushima**
## PR description
This PR fix the eth_getProof RPC method when using Bonsai by adding the worldStateProof to BonsaiWorldStateProvider.
By passing a snapshot of the ws that we want the proof for the BonsaiWorldStateProvider should be able to make normal usage of the worldStateProof to generate the proofs.

## Fixed Issue(s)

Part of #5377

### CODE REVIEW

**Review by matkt — CHANGES_REQUESTED** [review]
seems to be good

do we can add test for bonsai in this class WorldStateProofProviderTest.  the current tests are only using forest

you can use this class as example SnapWorldDownloadStateTest we are using Parameterized to call the same tests with bonsai and forest

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/proof/WorldStateProofProviderTest.java [hunk 1]
```diff
@@ -39,21 +44,43 @@
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
-import org.mockito.junit.MockitoJUnitRunner;
+import org.junit.runners.Parameterized;
 
-@RunWith(MockitoJUnitRunner.class)
+@RunWith(Parameterized.class)
```
Comment (siladu) [inline ethereum/core/src/test/java/org/hyperledger/besu/ethereum/proof/WorldStateProofProviderTest.java:49]: nit: would be preferable to use JUnit 5 syntax if possible

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/proof/WorldStateProofProviderTest.java [hunk 2]
```diff
@@ -39,21 +44,43 @@
...
   @Before
   public void setup() {
+    if (dataStorageFormat.equals(DataStorageFormat.BONSAI)) {
+      worldStateStorage =
+          new BonsaiWorldStateKeyValueStorage(
+              new InMemoryKeyValueStorageProvider(), new NoOpMetricsSystem());
+    } else {
+      worldStateStorage = new WorldStateKeyValueStorage(new InMemoryKeyValueStorage());
+    }
+
```
Comment (siladu) [inline ethereum/core/src/test/java/org/hyperledger/besu/ethereum/proof/WorldStateProofProviderTest.java:83]:
```suggestion
    worldStateStorage =
        switch (dataStorageFormat) {
          case BONSAI -> new BonsaiWorldStateKeyValueStorage(
              new InMemoryKeyValueStorageProvider(), new NoOpMetricsSystem());
          case FOREST -> new WorldStateKeyValueStorage(new InMemoryKeyValueStorage());
        };
```

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/proof/WorldStateProofProviderTest.java [hunk 3]
```diff
@@ -88,8 +115,13 @@ public void getProofWhenWorldStateAvailable() {
...
     worldStateTrie.put(addressHash, RLP.encode(accountValue::writeTo));
-    worldStateTrie.commit(updater::putAccountStateTrieNode);
-
+    worldStateTrie.commit(
+        (location, hash, value) -> {
+          updater.putAccountStateTrieNode(location, hash, value);
+          if (location.isEmpty()) {
+            updater.saveWorldState(Bytes.EMPTY, worldStateTrie.getRootHash(), value);
+          }
+        });
```
Comment (siladu) [inline ethereum/core/src/test/java/org/hyperledger/besu/ethereum/proof/WorldStateProofProviderTest.java:124]: Why is this necessary?

**Comment by matkt** [comment]
can we verify this GetProof method (compare with other client ?) from what i see getProof doesn't look good. I worked on this part a long time ago but I have a doubt now

Normally GetProof must return the siblings at each level in order to validate the proof. That doesn't seem like it regarding the code. I implemented a getProof method in another component so we can fix that easily but just wanted to be sure i'm not missing something

IMO it should be L1+L2+X for X

*         o
*        / \
*       o   o (L1)
*       / \
*  (L2)o  x

looking at the code it is L1+L2+X

* (L1)  o
*        / \
* (L2)o   o 
*       / \
*    o  x

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/proof/WorldStateProofProviderTest.java (hunk 3 above)
Comment (siladu) [reply ethereum/core/src/test/java/org/hyperledger/besu/ethereum/proof/WorldStateProofProviderTest.java:114]: Maybe worth a comment explaining this?

**Comment by gfukushima** [comment]
I've tested and we do provide the right proofs but there's seems to be an issue for proofs there are not the latest state in bonsai. Will move this PR to draft until I get back to it.

**Comment by gfukushima** [comment]
Closing this until I get back to it.

**Pull request closed without merging**